### ISSUE TITLE
Introduce Prettier to js files

### ISSUE THREAD

**Issue by jumbosushi** [ubyssey/ubyssey-dispatch-theme#621]
## Problem :skull:

- Our code base is unique in a sense that we get lot of contributors who are new to coding in general. Having their style mistakes pointed out on PR probably kills the momentum volunteers have.
- Checking style on PR from volunteer is also time-consuming.
- We don't have a style guidelines right now, which adds another thing to check / debate on PR

## Solution :zap:
Use [Prettier](https://prettier.io/) & [ESLint](https://eslint.org/) :tada: [This video at ReactConf](https://www.futurehosting.com/blog/prettier-vs-eslint-whats-the-difference/) is great intro for Prettier (only 8 min).

[This article](https://www.futurehosting.com/blog/prettier-vs-eslint-whats-the-difference/) compares the two pretty well.
TLDR is that people seem to use Prettier for auto-formatting (for both **js and css**!), and ESLint for style check (ex. typo on variable name, undeclared variable being used, etc). I think it'll be good for our repo to use both. If it's successful, we can copy the config to dispatch.

Prettier have an option to [add prettier as part of pre-commit hook](https://prettier.io/docs/en/precommit.html). This way, volunteers and devs don't have to even think about styling as it will be done automatically. By adding ESLint to this flow, we can even prevent bad code to be committed in the first place.

By doing this we can achieve:
- Consistent style across volunteers and devs
- No need to point out styling diffs on PR and focus on implementation instead
- Can be used as style guide

## Issues :triangular_flag_on_post:
I have couple thoughts on how to go about this, and wanna get input from you guys.

### 1. How much refactoring to do at once?

We have tons of js an css files. We can either make 1 PR with all files fixed, or fix files incrementally as people come across individual files.

I think we should go for incremental approach so we can manually glance over formatting as we go to make sure nothing breaks.

### 2. Change Git history?

Our existing git history is valuable, especially when we use it with `git blame` to see when the file was last edited / who to contact for q. Making a PR that formats files will over write this history with new commit history that just shows whoever ran the formatter, so it's not ideal.

I found couple article people overwriting git history for new formatter they introduced:
  - http://elliot.land/post/reformatting-your-codebase-with-git-filter-branch
  - https://medium.com/millennial-falcon-technology/reformatting-your-code-base-using-prettier-or-eslint-without-destroying-git-history-35052f3d853e
  - https://flexport.engineering/upgrading-prettier-on-a-large-codebase-28d56c4de49e

Although it's bit risky, I think we should go for overwriting past git history. I'm running one of the commands mentioned in the article to see how it looks locally, so I'll post the update later.

### 4. Both ESLint & Prettier?
Both tools are new to us, so it might make sense to use Prettier for a while before we start messing with ESLint.

From the config files I've looked over, it doesn't seem to bad. Think it's doable to do both at once.

### 5. How to introduce ESLint
   It's most likely our js files will fire up tons of ESLint like below:
![image](https://user-images.githubusercontent.com/9669739/50049353-c1534680-0098-11e9-901a-48f2733fc5e2.png)

I read [an article](https://tech.mercari.com/entry/2017/07/31/170125) (in jap) on how this one team did it. Apparently, they disabled all the config initially, then make individual PR to fix each lint. 
![image](https://user-images.githubusercontent.com/9669739/50049376-64a45b80-0099-11e9-8d5c-df867e5d305f.png)

I really like this approach in a sense that it keeps lint introduction more gradual, and also ** potentially create good intro issues for new volunteers**. 

### 6. Auto run formatter on commit or manual? (ex. `yarn prettier`)

React team at Facebook seem to enforce [a command to run prettier]() instead of doing it automatically. This is [the discussion they had when they introduced Prettier to React codebase](https://github.com/facebook/react/pull/9101#issuecomment-284034577)

Don't wanna add extra step for volunteers, so I feel like setting up pre-commit makes more sense.

Let me know what you think!

**Comment by jumbosushi** [comment]
@psiemens

**Comment by psiemens** [comment]
> Our code base is unique in a sense that we get lot of contributors who are new to coding in general. Having their style mistakes pointed out on PR probably kills the momentum volunteers have.

Such a good point! I really don't like having to point out small stylistic mistakes because I fear that it will discourage new contributors. This alone is enough motivation to use a linter.

> We have tons of js an css files. We can either make 1 PR with all files fixed, or fix files incrementally as people come across individual files.

ESLint is actually pretty good at automatically fixing most formatting mistakes. I think I'd opt for one PR that adds Prettier and ESLint and includes all the initial fixes - that way you get a nice clean codebase to work on in 2019 😄

> Prettier have an option to add prettier as part of pre-commit hook.

I didn't know that! Super cool.

> Our existing git history is valuable, especially when we use it with git blame to see when the file was last edited / who to contact for q. 

Also, with so many contributors I think it's really cool to see where everybody contributed. I've done some git rewrites in order to preserve the history before - don't have any problem with doing it again.

> Don't wanna add extra step for volunteers, so I feel like setting up pre-commit makes more sense.

I agree - that seems like overkill for now.

Thanks for proposing this @jumbosushi! Great idea 🎉

**Comment by brittkhat** [comment]
Investigating tools for 2022
